pkg/services/provisioning/dashboards/validator.go

```diff
@@ -0,0 +1,146 @@
+package dashboards
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+type duplicate struct {
+	Sum             uint8
+	InvolvedReaders map[string]struct{}
+}
+
+func newDuplicate() *duplicate {
+	return &duplicate{InvolvedReaders: make(map[string]struct{})}
+}
+
+type duplicateEntries struct {
+	Titles map[dashboardIdentity]*duplicate
+	UIDs   map[string]*duplicate
+}
+
+func (d *duplicateEntries) InvolvedReaders() map[string]struct{} {
+	involvedReaders := make(map[string]struct{})
+
+	for _, duplicate := range d.UIDs {
+		if duplicate.Sum <= 1 {
+			continue
+		}
+
+		for readerName := range duplicate.InvolvedReaders {
+			if _, ok := involvedReaders[readerName]; !ok {
+				involvedReaders[readerName] = struct{}{}
+			}
+		}
+	}
+
+	for _, duplicate := range d.Titles {
+		if duplicate.Sum <= 1 {
+			continue
+		}
+
+		for readerName := range duplicate.InvolvedReaders {
+			if _, ok := involvedReaders[readerName]; !ok {
+				involvedReaders[readerName] = struct{}{}
+			}
```

Comment:
There shouldn't be any practical benefit to only set this key if it's not already been set, so this will read more easily:
```suggestion
			involvedReaders[readerName] = struct{}{}
```